README.md
# resize.to options parser

[![Greenkeeper badge](https://badges.greenkeeper.io/resizeto/options-parser.svg)](https://greenkeeper.io/)
[![Coverage Status](https://coveralls.io/repos/github/resizeto/options-parser/badge.svg?branch=master)](https://coveralls.io/github/resizeto/options-parser?branch=master)

Parses a URI fragment and returns a collection of options to pass along to a [transformer](https://github.com/resizeto/transformer). This is a building block for creating image transformation services.

Example:

```
const uriFragment = 'width:100,height:100,output:webp/path/to/image.jpg'
const parser = new Parser(uriFragment)
const optionsCollection = parser.parse()
// optionsCollection == [{ width: 100, height: 100, output: 'webp' }]
// parser.path == 'path/to/image.jpg'
```

## URI Transform Options

## Options

* `backgroundalpha`: Default is `1`. Possible values are 0-1. This applies when cropping an image which results in the original image not completely covering the new dimensions. Example: `backgroundalpha:0.5`
* `backgroundcolor`: Default is `000`. This is the hex color code value excluding the `#`. This is parsed by the [color](https://www.npmjs.com/package/color) module. This applies when cropping an image which results in the original image not completely covering the new dimensions. Example: `backgroundcolor:000`
* `blur`: Possible values are 0.3-1000. Example: `blur:0.5`
* `fit`: Default is `cover`. Possible values are: `cover`, `contain`, `fill`, `inside`, and `outside`. Check the [sharp docs](https://sharp.pixelplumbing.com/en/stable/api-resize/) for more information. Example: `fit:cover`
* `gravity`: Default is `center`. Possible values are: `north`, `northeast`, `east`, `southeast`, `south`, `southwest`, `west`, `northwest`, `center`, `entropy`, `attention`. Check the [sharp docs](https://sharp.pixelplumbing.com/en/stable/api-resize/) for more information. Example: `gravity:attention`
* `height`: Default is to respect ratio. Acceptable values are a positive integer. Example: `height:100`
* `left`: This can be a float. Used for custom extraction/crop from the top left corner. Example: `width:100,height:100,left:100,top:100`
* `mirror`: Possible values are: `x` or `y`. Will mirror the image across the given axis. Example: `mirror:x`
* `output`: Default is to use the extension of the image. Possible values are: `jpeg`, `png`, or `webp`. Example: `output:webp`
* `quality`: Default is 80. Possible values are 1-100. Applies to `jpeg` and `webp`. Example: `quality:50`
* `rotate`: The value should be an angle and can be a float. The image is auto rotated based on exif data before applying other transforms. Example: `rotate:45`
* `signature`: Not required but recommended. A sha1 hex digest of the options (excluding signature) and path to the image. This will be verified on the server before processing.
* `top`: This can be a float. Used for custom extraction/crop from the top left corner. Example: `width:100,height:100,left:100,top:100`
* `version`: This is primarily for cache busting. Example: `version:101`
* `width`: Default is to respect ratio. Acceptable values are a positive integer. Example: `width:100`

## Compound Transforms

It is possible to do to multiple transformations serially by separating groups of options with a semi-colon. For example we could extract part of an image and then resize.

```
/top:20,left:5,width:250,height:250-width:100,height:100,gravity:entropy-signature:sha1hexdigest/path/to/image.jpg`
```

## Delimiters

There are a few options for what delimiters you can use out of the box. Here is a break down of the URI fragment that is relevant.

* Key-Value pairs (`width:100`): These can be separated by `_`, `:`, or `=`.
* Options or groups of key-value pairs (`width:100,height:100`): These can only be separated by `,`.
* Groups of options (`width:100-blur:0.3`): These can only be separated by `-`.


## Signing

It is highly recommended that you sign your URIs and keep your token private. This provides a level of protection against unauthorized transforms. Here is an example of generating a signed URI:

```
const crypto = require('crypto')

const secretToken = 'my-secret-token'

function sign (token, options, path) {
  const hmac = crypto.createHmac('sha1', token)
  hmac.update(`${options}/${path}`)
  const signature = hmac.digest('hex')
  return `${options}-signature:${signature}/${path}`
}

const signedFragment = sign(secretToken, 'width:100', 'path/to/image.jpg')
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.5]

### Changed

- Breaking: The compound/group delimiter was changed from `;` to `-`. The `;` was causing trouble for the S3 404 redirect solution where there isn't a way to URL encode the string to be passed as a query param.

## [0.0.4]

### Added

- Added name to error objects

## [0.0.3]

### Added

- Added `Options#hasOption(optionName)` method

### Changed

- The `Parser#optionsCollection` is a collection of frozen objects rather than an instance of the `Options` class
- Internally using `Options#_options` in Options class
- `Options#options` now returns a frozen object
- `Options#defaults` now returns a frozen object
- Renamed `Options#updateBackgroundRGB` to `Options#_updateBackgroundRGB`
- Limited files being included in npm package
- Updated dependencies

## [0.0.2]

### Added

- Added getters for Options


src/parser.test.js
/* eslint-env jest */
const { OptionKeyUnknownError, OptionValueInvalidError, SignatureMismatchError, SignatureMissingError } = require('./errors.js')
const Parser = require('./parser.js')
const crypto = require('crypto')

function sign (token, options, path) {
  const hmac = crypto.createHmac('sha1', token)
  hmac.update(`${options}/${path}`)
  const signature = hmac.digest('hex')
  return `${options}-signature:${signature}/${path}`
}

const path = 'path/to/image.jpg'

test('path works with leading /', () => {
  const options = 'width:100'
  const parser = new Parser(`/${options}/${path}`)
  parser.parse()
  expect(parser.path).toBe(path)
  expect(parser.optionsString).toBe(options)
})

test('path works without leading /', () => {
  const options = 'width:100'
  const parser = new Parser(`${options}/${path}`)
  parser.parse()
  expect(parser.path).toBe(path)
  expect(parser.optionsString).toBe(options)
})

test('signature missing', () => {
  const parser = new Parser(`width:100/${path}`, '123', true)
  expect(parser.token).toBe('123')
  expect(parser.signatureRequired).toBe(true)
  expect(() => { parser.parse() }).toThrow(SignatureMissingError)
  expect(() => { parser.verifySignature() }).toThrow(SignatureMismatchError)
})

test('signature mismatch', () => {
  const uri = sign('456', 'width:100', path)
  const parser = new Parser(uri, '123', true)
  expect(() => { parser.parse() }).not.toThrow(SignatureMissingError)
  expect(() => { parser.parse() }).toThrow(SignatureMismatchError)
})

test('signature match', () => {
  const uri = sign('123', 'width:100', path)
  const parser = new Parser(uri, '123', true)
  expect(() => { parser.parse() }).not.toThrow(SignatureMissingError)
  expect(() => { parser.parse() }).not.toThrow(SignatureMismatchError)
})

test('unknown option key', () => {
  const parser = new Parser(`notakey:notavalue/${path}`)
  expect(() => { parser.parse() }).toThrow(OptionKeyUnknownError)
})

test('invalid option value', () => {
  const parser = new Parser(`blur:0/${path}`)
  expect(() => { parser.parse() }).toThrow(OptionValueInvalidError)
})

test('optionsCollection is a collection of frozen objects', () => {
  const parser = new Parser(`width:100-blur:0.3/${path}`)
  const parsedOptionsCollection = parser.parse()
  expect(parsedOptionsCollection.length).toBe(2)
  expect(Object.isFrozen(parsedOptionsCollection[0])).toBe(true)
  expect(Object.isFrozen(parsedOptionsCollection[1])).toBe(true)
})

test('multiple options without signature', () => {
  const options = [
    ['width', '100'],
    ['height', '100'],
    ['output', 'webp']
  ]
  const expected = [
    ['width', 100],
    ['height', 100],
    ['output', 'webp']
  ]

  const delimiters = ['_', ':', '=']

  delimiters.forEach((delimiter) => {
    const optionsString = options.map((option) => option.join(delimiter)).join(',')

    const uri = `${optionsString}/${path}`
    const parser = new Parser(uri)
    const parsedOptionsCollection = parser.parse()[0]

    expected.forEach((expectedSetting) => {
      expect(parsedOptionsCollection[expectedSetting[0]]).toBe(expectedSetting[1])
    })
  })
})

test('multiple options with signature', () => {
  const options = [
    ['width', '100'],
    ['height', '100'],
    ['output', 'webp']
  ]
  const expected = [
    ['width', 100],
    ['height', 100],
    ['output', 'webp']
  ]

  const delimiters = ['_', ':', '=']

  delimiters.forEach((delimiter) => {
    const optionsString = options.map((option) => option.join(delimiter)).join(',')

    const uri = sign('123', optionsString, path)
    const parser = new Parser(uri, '123', true)
    const parsedOptionsCollection = parser.parse()[0]

    expected.forEach((expectedSetting) => {
      expect(parsedOptionsCollection[expectedSetting[0]]).toBe(expectedSetting[1])
    })
  })
})

test('compound options without signature', () => {
  const optionsCollection = [
    [
      ['width', '100'],
      ['height', '100']
    ],
    [
      ['blur', '0.3'],
      ['output', 'webp']
    ]
  ]
  const expectedCollection = [
    [
      ['width', 100],
      ['height', 100]
    ],
    [
      ['blur', 0.3],
      ['output', 'webp']
    ]
  ]

  const delimiters = ['_', ':', '=']

  delimiters.forEach((delimiter) => {
    const options = optionsCollection.map((options) => {
      return options.map((option) => option.join('_')).join(',')
    }).join('-')

    const parser = new Parser(`${options}/${path}`)
    const parsedOptionsCollection = parser.parse()
    parsedOptionsCollection.forEach((parsedOptions, index) => {
      expectedCollection[index].forEach((expectedSetting) => {
        expect(parsedOptions[expectedSetting[0]]).toBe(expectedSetting[1])
      })
    })
  })
})

test('compound options with signature', () => {
  const optionsCollection = [
    [
      ['width', '100'],
      ['height', '100']
    ],
    [
      ['blur', '0.3'],
      ['output', 'webp']
    ]
  ]
  const expectedCollection = [
    [
      ['width', 100],
      ['height', 100]
    ],
    [
      ['blur', 0.3],
      ['output', 'webp']
    ]
  ]

  const delimiters = ['_', ':', '=']

  delimiters.forEach((delimiter) => {
    const options = optionsCollection.map((options) => {
      return options.map((option) => option.join('_')).join(',')
    }).join('-')

    const uri = sign('123', options, path)
    const parser = new Parser(uri, '123', true)
    const parsedOptionsCollection = parser.parse()
    parsedOptionsCollection.forEach((parsedOptions, index) => {
      expectedCollection[index].forEach((expectedSetting) => {
        expect(parsedOptions[expectedSetting[0]]).toBe(expectedSetting[1])
      })
    })
  })
})


src/options.test.js
/* eslint-env jest */
const { OptionValueInvalidError } = require('./errors.js')
const Options = require('./options.js')

test('backgroundalpha', () => {
  const options = new Options()
  const values = ['1', '0', '0.5']
  const expected = [1.0, 0.0, 0.5]
  const expectedRGB = [
    'rgb(0, 0, 0)',
    'rgba(0, 0, 0, 0)',
    'rgba(0, 0, 0, 0.5)'
  ]
  expect(options.options.backgroundalpha).toBe(1)
  expect(options.backgroundalpha).toBe(1)
  expect(options.backgroundalpha).toBe(1)
  values.forEach((val, index) => {
    options.backgroundalpha = val
    expect(options.options.backgroundalpha).toBe(expected[index])
    expect(options.backgroundalpha).toBe(expected[index])
    expect(options.options.backgroundrgb).toBe(expectedRGB[index])
    expect(options.backgroundrgb).toBe(expectedRGB[index])
  })
})

test('backgroundcolor', () => {
  const options = new Options()
  const values = ['fff', 'ffffff']
  expect(options.options.backgroundcolor).toBe('#000')
  expect(options.backgroundcolor).toBe('#000')
  values.forEach((val) => {
    options.backgroundcolor = val
    expect(options.options.backgroundcolor).toBe(`#${val}`)
    expect(options.backgroundcolor).toBe(`#${val}`)
    expect(options.options.backgroundrgb).toBe('rgb(255, 255, 255)')
    expect(options.backgroundrgb).toBe('rgb(255, 255, 255)')
  })
})

test('blur', () => {
  const options = new Options()
  const values = ['0.3', '100', '1000']
  const expected = [0.3, 100, 1000]
  expect(options.options.blur).toBe(null)
  expect(() => { options.blur = '0' }).toThrow(OptionValueInvalidError)
  expect(() => { options.blur = '0.2' }).toThrow(OptionValueInvalidError)
  expect(() => { options.blur = '1001' }).toThrow(OptionValueInvalidError)
  values.forEach((val, index) => {
    options.blur = val
    expect(options.options.blur).toBe(expected[index])
    expect(options.blur).toBe(expected[index])
  })
})

test('fit', () => {
  const options = new Options()
  const values = ['cover', 'contain', 'fill', 'inside', 'outside']
  expect(options.options.fit).toBe('cover')
  expect(() => { options.fit = 'notvalid' }).toThrow(OptionValueInvalidError)
  values.forEach((val) => {
    options.fit = val
    expect(options.options.fit).toBe(val)
    expect(options.fit).toBe(val)
  })
})

test('gravity', () => {
  const options = new Options()
  const values = ['north', 'northeast', 'east', 'southeast', 'south', 'southwest', 'west', 'northwest', 'center', 'entropy', 'attention']
  expect(options.options.gravity).toBe('center')
  expect(() => { options.gravity = 'notvalid' }).toThrow(OptionValueInvalidError)
  values.forEach((val) => {
    options.gravity = val
    expect(options.options.gravity).toBe(val)
    expect(options.gravity).toBe(val)
  })
})

test('height', () => {
  const options = new Options()
  const values = ['1', '100.25', '1000']
  const expected = [1.0, 100.25, 1000]
  expect(options.options.height).toBe(null)
  values.forEach((val, index) => {
    options.height = val
    expect(options.options.height).toBe(expected[index])
    expect(options.height).toBe(expected[index])
  })
})

test('left', () => {
  const options = new Options()
  const values = ['1', '100.25', '1000']
  const expected = [1.0, 100.25, 1000]
  expect(options.options.left).toBe(null)
  values.forEach((val, index) => {
    options.left = val
    expect(options.options.left).toBe(expected[index])
    expect(options.left).toBe(expected[index])
  })
})

test('mirror', () => {
  const options = new Options()
  const values = ['x', 'y']
  expect(options.options.mirror).toBe(null)
  expect(() => { options.mirror = 'a' }).toThrow(OptionValueInvalidError)
  values.forEach((val) => {
    options.mirror = val
    expect(options.options.mirror).toBe(val)
    expect(options.mirror).toBe(val)
  })
})

test('output', () => {
  const options = new Options()
  const values = ['jpeg', 'png', 'webp']
  expect(options.options.output).toBe(null)
  expect(() => { options.output = 'invalid' }).toThrow(OptionValueInvalidError)
  values.forEach((val) => {
    options.output = val
    expect(options.options.output).toBe(val)
    expect(options.output).toBe(val)
  })
})

test('quality', () => {
  const options = new Options()
  const values = ['1', '100', '50']
  const expected = [1, 100, 50]
  expect(options.options.quality).toBe(80)
  expect(() => { options.quality = '0' }).toThrow(OptionValueInvalidError)
  expect(() => { options.quality = '101' }).toThrow(OptionValueInvalidError)
  values.forEach((val, index) => {
    options.quality = val
    expect(options.options.quality).toBe(expected[index])
    expect(options.quality).toBe(expected[index])
  })
})

test('rotate', () => {
  const options = new Options()
  const values = ['0', '-450', '10.5']
  const expected = [0, -450.0, 10.5]
  expect(options.options.rotate).toBe(null)
  values.forEach((val, index) => {
    options.rotate = val
    expect(options.options.rotate).toBe(expected[index])
    expect(options.rotate).toBe(expected[index])
  })
})

test('top', () => {
  const options = new Options()
  const values = ['1', '100.25', '1000']
  const expected = [1.0, 100.25, 1000]
  expect(options.options.top).toBe(null)
  values.forEach((val, index) => {
    options.top = val
    expect(options.options.top).toBe(expected[index])
    expect(options.top).toBe(expected[index])
  })
})

test('version', () => {
  const options = new Options()
  const values = ['test', '3']
  expect(options.options.version).toBe(null)
  values.forEach((val) => {
    options.version = val
    expect(options.options.version).toBe(val)
    expect(options.version).toBe(val)
  })
})

test('width', () => {
  const options = new Options()
  const values = ['1', '100.25', '1000']
  const expected = [1.0, 100.25, 1000]
  expect(options.options.width).toBe(null)
  values.forEach((val, index) => {
    options.width = val
    expect(options.options.width).toBe(expected[index])
    expect(options.width).toBe(expected[index])
  })
})

test('defaults are frozen', () => {
  const options = new Options()
  expect(Object.isFrozen(options.defaults)).toBe(true)
})

test('options are frozen', () => {
  const options = new Options()
  expect(Object.isFrozen(options.options)).toBe(true)
})

test('hasOption', () => {
  const options = new Options()
  expect(options.hasOption('gravity')).toBe(true)
  expect(options.hasOption('unknown')).toBe(false)
})

test('extendable', () => {
  class MyOptions extends Options {
    get defaults () {
      return Object.freeze(Object.assign({}, super.defaults, {
        text: '',
        gravity: 'special'
      }))
    }

    set text (value) {
      this._options.text = value
    }

    set gravity (value) {
      const valid = ['north', 'northeast', 'east', 'southeast', 'south', 'southwest', 'west', 'northwest', 'center', 'entropy', 'attention', 'special']
      if (valid.indexOf(value) === -1) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'gravity'`)
      this._options.gravity = value
    }
  }

  const options = new MyOptions()
  const textValues = ['test', '3']
  const gravityValues = ['north', 'special']

  expect(options.hasOption('text')).toBe(true)
  expect(options.options.text).toBe('')
  expect(options.text).toBe('')
  textValues.forEach((val) => {
    options.text = val
    expect(options.options.text).toBe(val)
    expect(options.text).toBe(val)
  })

  expect(options.options.gravity).toBe('special')
  expect(options.gravity).toBe('special')
  gravityValues.forEach((val) => {
    options.gravity = val
    expect(options.options.gravity).toBe(val)
    expect(options.gravity).toBe(val)
  })
})


src/parser.js
const { OptionKeyUnknownError, SignatureMismatchError, SignatureMissingError } = require('./errors.js')
const Options = require('./options.js')
const crypto = require('crypto')

class Parser {
  constructor (uri, token, signatureRequired) {
    this.uri = uri.replace(/^\//, '')
    this.token = token
    this.signatureRequired = signatureRequired
    this.optionDelimiters = /[,]/
    this.compoundDelimiters = /[-]/
    this.keyValueDelimiters = /[_:=]/
    this.signatureRegExp = /[,-]signature[_:=]([\w\d]+)/
    this.optionsCollection = []
  }

  uriComponents () {
    const components = this.uri.split('/')
    this.optionsString = components.shift()
    const signatureMatch = this.optionsString.match(this.signatureRegExp)
    if (signatureMatch) {
      this.signature = signatureMatch[1]
      this.optionsString = this.optionsString.replace(this.signatureRegExp, '')
    }
    this.path = components.join('/')
  }

  parse () {
    this.uriComponents()

    if (this.signature) this.verifySignature()
    else if (this.signatureRequired) throw new SignatureMissingError('Signature not found')

    this.optionsCollection = this.optionsString.split(this.compoundDelimiters).map((optionsString) => {
      const options = new Options()
      optionsString.split(this.optionDelimiters).map((optionString) => {
        const [key, value] = optionString.split(this.keyValueDelimiters)
        if (!options.hasOption(key)) throw new OptionKeyUnknownError(`${key} is not a known option`)
        options[key] = value
      })
      return options.options
    })

    return this.optionsCollection
  }

  verifySignature () {
    const hmac = crypto.createHmac('sha1', this.token)
    hmac.update(`${this.optionsString}/${this.path}`)
    const signature = hmac.digest('hex')
    if (this.signature !== signature) throw new SignatureMismatchError('Signature did not match')
  }
}

module.exports = Parser


src/options.js
const Color = require('color')
const { OptionValueInvalidError } = require('./errors.js')

// All setters expect to get a string
class Options {
  constructor () {
    this._options = Object.assign({}, this.defaults)

    // We'll use the keys from the defaults
    // to proxy get calls directly to the
    // _options object
    const keys = Object.keys(this.defaults)

    return new Proxy(this, {
      get (target, name) {
        if (keys.indexOf(name) !== -1) {
          return Reflect.get(target, '_options')[name]
        } else {
          return target[name]
        }
      },

      set (target, name, value) {
        return Reflect.set(...arguments)
      }
    })
  }

  hasOption (key) {
    return key in this._options
  }

  _updateBackgroundRGB () {
    const color = Color(this._options.backgroundcolor).alpha(this._options.backgroundalpha)
    this._options.backgroundrgb = color.rgb().string()
  }

  get options () {
    return Object.freeze(Object.assign({}, this._options))
  }

  get defaults () {
    return Object.freeze({
      backgroundalpha: 1, // 0-1
      backgroundcolor: '#000', // hex without the #
      backgroundrgb: 'rgb(0, 0, 0)', // automatically updated when color or alpha is set
      blur: null, // 0.3-1000
      fit: 'cover', // cover, contain, fill, inside, outside
      gravity: 'center', // north, northeast, east, southeast, south, southwest, west, northwest, center, entropy, attention
      height: null, // auto/aspect ratio by default
      left: null, // number (used for custom extraction/crop)
      mirror: null, // x, y (which axis to mirror it across)
      output: null, // jpeg, png, webp
      quality: 80, // 1-100 applies to jpeg and webp
      rotate: null, //  always auto rotates based on exif
      top: null, // number (used for custom extraction/crop)
      version: null, // anything
      width: null // auto/aspect ratio by default
    })
  }

  set backgroundalpha (value) {
    this._options.backgroundalpha = parseFloat(value)
    this._updateBackgroundRGB()
  }

  set backgroundcolor (value) {
    this._options.backgroundcolor = `#${value}`
    this._updateBackgroundRGB()
  }

  set blur (value) {
    value = parseFloat(value)
    if (value < 0.3 || value > 1000) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'blur'`)
    this._options.blur = value
  }

  set fit (value) {
    const valid = ['cover', 'contain', 'fill', 'inside', 'outside']
    if (valid.indexOf(value) === -1) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'fit'`)
    this._options.fit = value
  }

  set gravity (value) {
    const valid = ['north', 'northeast', 'east', 'southeast', 'south', 'southwest', 'west', 'northwest', 'center', 'entropy', 'attention']
    if (valid.indexOf(value) === -1) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'gravity'`)
    this._options.gravity = value
  }

  set height (value) {
    this._options.height = parseFloat(value)
  }

  set left (value) {
    this._options.left = parseFloat(value)
  }

  set mirror (value) {
    const valid = ['x', 'y']
    if (valid.indexOf(value) === -1) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'mirror'`)
    this._options.mirror = value
  }

  set output (value) {
    const valid = ['jpeg', 'png', 'webp']
    if (valid.indexOf(value) === -1) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'output'`)
    this._options.output = value
  }

  set quality (value) {
    value = parseInt(value, 10)
    if (value < 1 || value > 100) throw new OptionValueInvalidError(`'${value}' is an invalid value for option 'quality'`)
    this._options.quality = value
  }

  set rotate (value) {
    this._options.rotate = parseFloat(value)
  }

  set top (value) {
    this._options.top = parseFloat(value)
  }

  set version (value) {
    this._options.version = value
  }

  set width (value) {
    this._options.width = parseFloat(value)
  }
}

module.exports = Options
